lib/elixir/src/elixir_expand.erl

```diff
@@ -272,6 +272,22 @@ expand({for, Meta, [_ | _] = Args}, E) ->
         form_error(Meta, ?key(E, file), ?MODULE, {missing_option, for, [do]})
     end,
 
+  IsReducing = lists:keymember(reduce, 1, Opts),
+  IsClauses = is_list(Expr) andalso lists:any(fun({'->', _, _}) -> true; (_) -> false end, Expr),
+
+  case IsReducing andalso (lists:keymember(into, 1, Opts) orelse lists:keymember(uniq, 1, Opts)) of
+    true ->
+      form_error(Meta, ?key(E, file), ?MODULE, for_duplicate_reduce_into_uniq);
+    false ->
+      ok
+  end,
+
+  case {IsReducing, IsClauses} of
+    {true, false} -> form_error(Meta, ?key(E, file), ?MODULE, for_with_reduce_bad_args);
```

Comment:
And maybe these two `for_with_reduce_bad_args -> for_with_reduce_bad_do_block` and `for_without_reduce_with_arg -> for_bad_do_block`?

Reply:
Since those names are only used internally, I think it is fine to keep them as is. :+1: